tiled/server/authentication.py

```diff
@@ -503,76 +526,570 @@ async def create_tokens_from_session(settings, db, session, provider):
-            "user_code": pending_session["user_code"],
-        }
-
-    return route
 
-
-def build_device_code_user_code_form_route(authentication, provider):
-    if not SHARE_TILED_PATH:
+    @authentication_router.delete(
+        "/principal/{uuid}/apikey",
+        response_model=schemas.Principal,
+        dependencies=[Security(lambda: None, scopes=["admin:apikeys"])],
+    )
+    async def revoke_apikey_for_principal(
+        request: Request,
+        uuid: uuid_module.UUID,
+        first_eight: str,
+        db=Depends(get_database_session),
+    ):
+        "Allow Tiled Admins to delete any user's apikeys e.g."
+        request.state.endpoint = "auth"
+        api_key_orm = (
+            await db.execute(
+                select(orm.APIKey).filter(orm.APIKey.first_eight == first_eight[:8])
+            )
+        ).scalar()
+        if (api_key_orm is None) or (api_key_orm.principal.uuid != uuid):
+            raise HTTPException(
+                404,
+                f"The principal {uuid} has no such API key.",
+            )
+        await db.delete(api_key_orm)
+        await db.commit()
+
+        return Response(status_code=HTTP_204_NO_CONTENT)
+
+    @authentication_router.post(
+        "/principal/{uuid}/apikey",
+        response_model=schemas.APIKeyWithSecret,
+    )
+    async def apikey_for_principal(
+        request: Request,
+        uuid: uuid_module.UUID,
+        apikey_params: schemas.APIKeyRequestParams,
+        principal=Security(get_current_principal, scopes=["admin:apikeys"]),
+        db=Depends(get_database_session),
+    ):
+        "Generate an API key for a Principal."
+        request.state.endpoint = "auth"
+        principal = (
+            await db.execute(select(orm.Principal).filter(orm.Principal.uuid == uuid))
+        ).scalar()
+        if principal is None:
+            raise HTTPException(
+                404, f"Principal {uuid} does not exist or insufficient permissions."
+            )
+        return await generate_apikey(db, principal, apikey_params, request)
+
+    @authentication_router.post(
+        "/session/refresh", response_model=schemas.AccessAndRefreshTokens
+    )
+    async def refresh_session(
+        request: Request,
+        refresh_token: schemas.RefreshToken,
+        settings: Settings = Depends(get_settings),
+        db=Depends(get_database_session),
+    ):
+        "Obtain a new access token and refresh token."
+        request.state.endpoint = "auth"
+        new_tokens = await slide_session(refresh_token.refresh_token, settings, db)
+        return new_tokens
+
+    @authentication_router.post("/session/revoke")
+    async def revoke_session(
+        request: Request,
+        refresh_token: schemas.RefreshToken,
+        db=Depends(get_database_session),
+    ):
+        "Mark a Session as revoked so it cannot be refreshed again."
+        request.state.endpoint = "auth"
+        payload = await decode_access_token(refresh_token.refresh_token)
```

Comment:
Something's crossed here, where we're decoding the refresh token with `decode_access_token`.

Reply:
Yeah, I think it should be named `decode_token` but otherwise it's solid: there's a "get access token from request" method and a "decode token" that can take an access or request token.